### PR TITLE
feature: show devices with benchmarks

### PR DESCRIPTION
**Pull request by sshaman1101**
This PR improves device list shown to Worker's owner.
The list includes new device groups (network, storage).
Also, benchmarks results for every device group is shown.

### CODE REVIEW

File: insonmnia/hardware/cpu/device.go [hunk 1]
```diff
@@ -1,27 +1,28 @@
...
 	if err != nil {
 		return nil, err
 	}
 
-	devices := []Device{}
-	for _, device := range info {
-		dev := Device(device)
-		devices = append(devices, dev)
+	if len(info) == 0 {
+		// o_O
```
Comment (3Hren) [inline insonmnia/hardware/cpu/device.go]: Remove

File: insonmnia/hardware/cpu/device.go [hunk 2]
```diff
@@ -1,27 +1,28 @@
...
+		return nil, errors.New("no CPU detected")
+	}
+
+	dev := &sonm.CPUDevice{ModelName: info[0].ModelName}
```
Comment (3Hren) [inline insonmnia/hardware/cpu/device.go]: And here the description is required, it's not clear why you take the first model name for the entire device set.

File: insonmnia/hardware/cpu/device.go [hunk 3]
```diff
@@ -1,27 +1,28 @@
...
+	for _, c := range info {
+		dev.Cores += uint32(c.Cores)
+		dev.Sockets += 1
```
Comment (3Hren) [inline insonmnia/hardware/cpu/device.go]: Really? `dev.Sockets = len(info)`

File: insonmnia/hardware/hardware.go [hunk 4]
```diff
@@ -9,122 +9,79 @@ import (
...
 // DeviceMapping maps hardware capabilities to device description, hashing-friendly
 type DeviceMapping struct {
-	CPU     []cpu.Device      `json:"cpu"`
-	GPU     []*sonm.GPUDevice `json:"gpu"`
-	Memory  HashableMemory    `json:"memory"`
-	Network interface{}       `json:"network"`
-	Storage interface{}       `json:"storage"`
+	CPU     *sonm.CPUDevice     `json:"cpu"`
+	GPU     []*sonm.GPUDevice   `json:"gpu"`
+	Memory  hashableMemory      `json:"memory"`
```
Comment (3Hren) [inline insonmnia/hardware/hardware.go]: So RAM or Memory? Let's be consistent.

File: insonmnia/hardware/hardware.go [hunk 5]
```diff
@@ -9,122 +9,79 @@ import (
...
-	for _, c := range h.CPU {
-		m.CPU = append(m.CPU, c.Device)
-	}
-
-	for _, g := range h.GPU {
-		m.GPU = append(m.GPU, g.Device)
+	return &DeviceMapping{
+		CPU:     h.CPU.Device,
+		GPU:     GPUs,
+		Memory:  hashableMemory{Available: h.RAM.Device.Available},
```
Comment (3Hren) [inline insonmnia/hardware/hardware.go]: Here too.

File: insonmnia/hardware/marshal.go [hunk 6]
```diff
@@ -5,36 +5,11 @@ import (
...
-
 	return &sonm.DevicesReply{
-		CPUs:    cpus,
-		GPUs:    gpus,
-		Memory:  ram,
-		Network: net,
-		Storage: storage,
+		CPU:     h.CPU,
+		GPUs:    h.GPU,
+		Memory:  h.RAM,
```
Comment (3Hren) [inline insonmnia/hardware/marshal.go]: image

File: insonmnia/miner/server.go [hunk 7]
```diff
@@ -803,27 +811,36 @@ func (m *Miner) runGPUBenchGroup(benches []*pb.Benchmark) error {
 	for _, dev := range m.hardware.GPU {
 		for _, ben := range benches {
 			if ben.GetID() == bm.GPUCount {
-				dev.Benchmark[bm.GPUCount] = &pb.Benchmark{ID: ben.GetID(), Result: 1}
+				dev.Benchmarks[bm.GPUCount] = &pb.Benchmark{
+					ID:     ben.GetID(),
+					Code:   ben.GetCode(),
+					Result: 1,
```
Comment (3Hren) [inline insonmnia/miner/server.go]: Stub?

File: proto/hub.proto [hunk 8]
```diff
@@ -98,11 +98,11 @@ message TaskListReply {
...
 
 message DevicesReply {
-    repeated CPUDevice CPUs = 1;
-    repeated GPUDevice GPUs = 2;
-    RAMDevice memory = 3;
-    NetworkDevice network = 4;
-    StorageDevice storage = 5;
+    CPU CPU = 1;
+    repeated GPU GPUs = 2;
+    RAM memory = 3;
```
Comment (3Hren) [inline proto/hub.proto]: Inconsistence.

File: proto/capabilities.proto [hunk 9]
```diff
@@ -70,18 +63,28 @@ message GPUDevice {
...
 }
 
 message StorageDevice {
     uint64 bytesAvailable = 1;
-    map<uint64, Benchmark> benchmarks = 2;
+}
+
+message Storage {
+    StorageDevice device = 1;
+    map<uint64, Benchmark> benchmarks = 4;
```
Comment (3Hren) [inline proto/capabilities.proto]: =4?

File: proto/capabilities.proto [hunk 10]
```diff
@@ -70,18 +63,28 @@ message GPUDevice {
...
 
 message NetworkDevice {
     uint64 bandwidthIn = 1;
     uint64 bandwidthOut = 2;
+}
 
-    map<uint64, Benchmark> benchmarks = 3;
+message Network {
+    NetworkDevice device = 1;
+    map<uint64, Benchmark> benchmarks = 4;
```
Comment (3Hren) [inline proto/capabilities.proto]: =4?

File: proto/capabilities.proto [hunk 11]
```diff
@@ -70,18 +63,28 @@ message GPUDevice {
 
     // Memory is amount of vmem for device, in bytes
     uint64 Memory = 9;
+}
 
-    map<uint64, Benchmark> benchmarks = 10;
+message GPU {
+    GPUDevice device = 1;
+    map<uint64, Benchmark> benchmarks = 4;
```
Comment (3Hren) [inline proto/capabilities.proto]: =4?

File: proto/capabilities.proto [hunk 12]
```diff
@@ -17,35 +17,28 @@ message Capabilities {
...
     uint64 total = 1;
-    // fixme: do we really need this?
-    uint64 used = 2;
+    uint64 available = 2;
+}
 
-    map<uint64, Benchmark> benchmarks = 3;
+message RAM {
+    RAMDevice device = 1;
+    map<uint64, Benchmark> benchmarks = 4;
```
Comment (3Hren) [inline proto/capabilities.proto]: =4?

File: proto/capabilities.proto [hunk 13]
```diff
@@ -17,35 +17,28 @@ message Capabilities {
...
-
-    map<uint64, Benchmark> benchmarks = 10;
+    uint32 cores = 2;
+    // Sockets describes number of CPU sockets on a host system.
+    uint32 sockets = 3;
+}
+
+message CPU {
+    CPUDevice device = 1;
+    map<uint64, Benchmark> benchmarks = 4;
```
Comment (3Hren) [inline proto/capabilities.proto]: =4?

File: insonmnia/hardware/cpu/device.go (hunk 1 above)
Comment (antmat) [inline insonmnia/hardware/cpu/device.go]: panic?

File: insonmnia/miner/server.go (hunk 7 above)
Comment (sshaman1101) [reply insonmnia/miner/server.go]: No, this benchmark measures an amount of GPUs for... hmm, one GPU unit.

**Pull request merged**